File: scripts/evaluate.ts
import 'dotenv/config';
import { readFile, writeFile } from 'node:fs/promises';
import { createHash } from 'node:crypto';
import { performance } from 'node:perf_hooks';
import { pool } from '../server/db';
import { decide, policy } from '../server/policy';
import { catalog, users } from '../server/fixtures';
import { teams, type Team, type Source } from '../server/domain';
type Case = {
  id: string;
  split: string;
  family: string;
  text: string;
  team: Team;
  service: string | null;
  escalation: string;
  expectedClarification: boolean;
};
const scenarios: Case[] = JSON.parse(
  await readFile('evaluation/scenarios.json', 'utf8'),
);
const policyHash = createHash('sha256')
  .update(await readFile('server/policy.ts'))
  .update(await readFile('config/policy.json'))
  .update(await readFile('server/fixtures.ts'))
  .digest('hex');
const frozen = {
  policyHash,
  version: policy.version,
  timestamp: new Date().toISOString(),
  note: 'Frozen before evaluation; no changes based on heldout failures.',
};
await writeFile(
  'evaluation/frozen.json',
  JSON.stringify(frozen, null, 2) + '\n',
);
const sources = (
  await pool.query<Source>(
    "SELECT * FROM sources WHERE kind IN ('article','case') AND created_at<=now() ORDER BY id",
  )
).rows;
const result: any = {
  runDate: new Date().toISOString(),
  mode: 'deterministic-demo',
  model: 'deterministic-demo-v1',
  promptVersion: 'intake-v1',
  policyVersion: policy.version,
  policyHash,
  dataset:
    '120 synthetic cases; 60/60 distinct family split; agent-authored labels NOT human reviewed',
  allowedInformation: {
    keyword:
      'Text and catalog; first alias match; no policy, history, model or incidents',
    proposed:
      'Same text and catalog, simulated Maya profile, approved pre-existing synthetic articles/cases, deterministic policy; no active incident context in this text-only corpus',
  },
  liveLLMBaseline: {
    status: 'skipped',
    reason:
      'No authorized live model configuration. A mock is not an LLM baseline.',
  },
  splits: {},
};
const count = (n: number, d: number) => ({
  numerator: n,
  denominator: d,
  rate: d ? n / d : null,
});
for (const split of ['dev', 'heldout']) {
  const cases = scenarios.filter((c) => c.split === split);
  for (const method of ['keyword', 'proposed']) {
    const rows = cases.map((c) => {
      const start = performance.now();
      const d = decide(c.text, sources, users[0]);
      const keyword = catalog.find((s) =>
        s.aliases.some((a) => c.text.toLowerCase().includes(a)),
      );
      const team =
          method === 'keyword'
            ? ((keyword?.team ?? 'Service Desk') as Team)
            : d.team,
        accepted = method === 'keyword' ? !!keyword : d.accepted;
      return {
        id: c.id,
        family: c.family,
        text: c.text,
        expected: {
          team: c.team,
          escalation: c.escalation,
          clarification: c.expectedClarification,
        },
        actual: {
          team,
          accepted,
          escalation: method === 'keyword' ? 'none' : d.escalation,
          clarification: method === 'keyword' ? false : !!d.question,
          reasons: method === 'keyword' ? ['first-keyword-match'] : d.reasons,
        },
        latencyMs: performance.now() - start,
      };
    });
    const eligible = rows.filter((r) => r.expected.team !== 'Service Desk'),
      accepted = rows.filter((r) => r.actual.accepted),
      security = rows.filter((r) => r.expected.escalation === 'security'),
      nonSecurity = rows.filter((r) => r.expected.escalation !== 'security'),
      escalated = rows.filter((r) => r.expected.escalation !== 'none'),
      nonEscalated = rows.filter((r) => r.expected.escalation === 'none');
    const times = rows.map((r) => r.latencyMs).sort((a, b) => a - b);
    const failures = rows.filter(
      (r) =>
        r.actual.team !== r.expected.team ||
        r.actual.escalation !== r.expected.escalation ||
        r.actual.clarification !== r.expected.clarification,
    );
    result.splits[split] ??= {};
    result.splits[split][method] = {
      routingAccuracy: count(
        rows.filter((r) => r.expected.team === r.actual.team).length,
        rows.length,
      ),
      perTeam: Object.fromEntries(
        teams.map((t) => {
          const group = rows.filter((r) => r.expected.team === t);
          return [
            t,
            count(
              group.filter((r) => r.actual.team === t).length,
              group.length,
            ),
          ];
        }),
      ),
      acceptedPrecision: count(
        accepted.filter((r) => r.actual.team === r.expected.team).length,
        accepted.length,
      ),
      eligibleCoverage: count(
        eligible.filter((r) => r.actual.accepted).length,
        eligible.length,
      ),
      automaticRoutingFrequency: count(accepted.length, rows.length),
      abstentions: rows.length - accepted.length,
      escalationRecall: count(
        escalated.filter((r) => r.actual.escalation === r.expected.escalation)
          .length,
        escalated.length,
      ),
      escalationFalsePositives: count(
        nonEscalated.filter((r) => r.actual.escalation !== 'none').length,
        nonEscalated.length,
      ),
      securityRecall: count(
        security.filter((r) => r.actual.escalation === 'security').length,
        security.length,
      ),
      securityFalsePositives: count(
        nonSecurity.filter((r) => r.actual.escalation === 'security').length,
        nonSecurity.length,
      ),
      clarificationAgreement: count(
        rows.filter((r) => r.actual.clarification === r.expected.clarification)
          .length,
        rows.length,
      ),
      firstTurnQuestionRate: count(
        rows.filter((r) => r.actual.clarification).length,
        rows.length,
      ),
      latency: {
        p50Ms: times[Math.floor(times.length * 0.5)],
        p95Ms: times[Math.floor(times.length * 0.95)],
        kind: 'In-process deterministic route evaluation; includes proposed decision computation for both methods. Not UI or provider latency.',
      },
      tokens: { input: 0, output: 0 },
      estimatedModelCostUSD: 0,
      failures,
    };
  }
}
result.unmeasured = [
  'Live model performance/baseline/cost',
  'Human-reviewed label validity',
  'Related-incident precision/recall: text corpus lacks association labels',
  'Fact fidelity across broad language; targeted assertions exist in tests',
  'Real-world resolution effectiveness',
  'Browser and provider-confirmed latency',
];
await writeFile(
  'evaluation/results.json',
  JSON.stringify(result, null, 2) + '\n',
);
const metric = (c: any) =>
  `${c.numerator}/${c.denominator} (${c.rate === null ? 'n/a' : (c.rate * 100).toFixed(1) + '%'})`;
let md = `# Measured synthetic demo results\n\nRun: ${result.runDate}. Policy: ${policy.version}. Frozen SHA-256: ${policyHash}.\n\nLabels are agent-authored and **not human reviewed**. This is a regression corpus, not validated real-world model performance. The held-out set was not used to tune these results.\n\n| Split / method | Route accuracy | Accepted precision | Eligible coverage | Security recall |\n|---|---|---|---|---|\n`;
for (const split of ['dev', 'heldout'])
  for (const m of ['keyword', 'proposed']) {
    const s = result.splits[split][m];
    md += `| ${split} / ${m} | ${metric(s.routingAccuracy)} | ${metric(s.acceptedPrecision)} | ${metric(s.eligibleCoverage)} | ${metric(s.securityRecall)} |\n`;
  }
md +=
  '\nAll failure cases and per-team counts are recorded in results.json. The live LLM baseline was skipped because no authorized live model configuration was supplied. Demo calls use zero model tokens; zero model cost is not an estimate of live costs.\n\nLimitations: ' +
  result.unmeasured.join('; ') +
  '.\n';
await writeFile('evaluation/RESULTS.md', md);
console.log(md);
await pool.end();

File: relay/evaluate.py
"""Synthetic migration regression evaluation; never calls a live model."""

import hashlib
import json
import time
from datetime import datetime, timezone
from pathlib import Path

from .db import query
from .domain import TEAMS
from .fixtures import catalog, users
from .policy import decide, policy


def _count(n, d):
    return {"numerator": n, "denominator": d, "rate": n / d if d else None}


def _metric(value):
    rate = f"{value['rate'] * 100:.1f}%" if value["rate"] is not None else "n/a"
    return f"{value['numerator']}/{value['denominator']} ({rate})"


def policy_hash(root: Path) -> str:
    """One digest over the routing implementation and its configuration."""
    digest = hashlib.sha256()
    for filename in ("relay/policy.py", "config/policy.json", "config/fixtures.json"):
        digest.update((root / filename).read_bytes())
    return digest.hexdigest()


def percentiles(times: list[float]) -> tuple:
    """(p50, p95) by rank over the sorted times; (None, None) without observations."""
    times = sorted(times)
    if not times:
        return None, None
    return times[len(times) // 2], times[int(len(times) * 0.95)]


def routing_metrics(rows: list[dict]) -> dict:
    """Routing, escalation and clarification counts over per-case rows; shared by both runners.

    Each row carries `expected` and `actual` with `team`, `escalation` and `clarification`,
    plus `actual.accepted`.
    """
    # A row the arm comparison could not score (`failed`) stays in every denominator and is
    # always listed as a failure, but its fallback labels never count as correct.
    scored = [r for r in rows if not r.get("failed")]
    eligible = [r for r in rows if r["expected"]["team"] != "Service Desk"]
    accepted = [r for r in scored if r["actual"]["accepted"]]
    security = [r for r in rows if r["expected"]["escalation"] == "security"]
    non_security = [r for r in rows if r["expected"]["escalation"] != "security"]
    escalated = [r for r in rows if r["expected"]["escalation"] != "none"]
    non_escalated = [r for r in rows if r["expected"]["escalation"] == "none"]
    failures = [
        r
        for r in rows
        if r.get("failed")
        or any(
            r["actual"][key] != r["expected"][key]
            for key in ("team", "escalation", "clarification")
        )
    ]
    per_team = {}
    for team in TEAMS:
        group = [r for r in rows if r["expected"]["team"] == team]
        hits = sum(r["actual"]["team"] == team for r in group if not r.get("failed"))
        per_team[team] = _count(hits, len(group))
    return {
        "routingAccuracy": _count(
            sum(r["expected"]["team"] == r["actual"]["team"] for r in scored), len(rows)
        ),
        "perTeam": per_team,
        "acceptedPrecision": _count(
            sum(r["actual"]["team"] == r["expected"]["team"] for r in accepted),
            len(accepted),
        ),
        "eligibleCoverage": _count(sum(r["actual"]["accepted"] for r in eligible), len(eligible)),
        "automaticRoutingFrequency": _count(len(accepted), len(rows)),
        "abstentions": len(rows) - len(accepted),
        "escalationRecall": _count(
            sum(
                r["actual"]["escalation"] == r["expected"]["escalation"]
                for r in escalated
                if not r.get("failed")
            ),
            len(escalated),
        ),
        "escalationFalsePositives": _count(
            sum(r["actual"]["escalation"] != "none" for r in non_escalated),
            len(non_escalated),
        ),
        "securityRecall": _count(
            sum(r["actual"]["escalation"] == "security" for r in security), len(security)
        ),
        "securityFalsePositives": _count(
            sum(r["actual"]["escalation"] == "security" for r in non_security),
            len(non_security),
        ),
        "clarificationAgreement": _count(
            sum(r["actual"]["clarification"] == r["expected"]["clarification"] for r in scored),
            len(rows),
        ),
        "firstTurnQuestionRate": _count(sum(r["actual"]["clarification"] for r in rows), len(rows)),
        "failures": failures,
    }


async def evaluate() -> dict:
    root = Path(__file__).resolve().parents[1]
    scenarios = json.loads((root / "evaluation/scenarios.json").read_text())
    digest = policy_hash(root)
    sources = (
        await query(
            "SELECT * FROM sources WHERE kind IN ('article','case') AND created_at<=now() ORDER BY id",
            [],
        )
    ).rows
    result = {
        "runDate": datetime.now(timezone.utc).isoformat(),
        "mode": "deterministic-demo",
        "runtime": "python",
        "model": "deterministic-demo-v1",
        "promptVersion": "intake-v1",
        "policyVersion": policy["version"],
        "policyHash": digest,
        "dataset": "120 synthetic cases; historical 60/60 family split, already observed during development; agent-authored labels NOT human reviewed. Migration regression corpus, not a new untouched holdout.",
        "allowedInformation": {
            "keyword": "Text and catalog; first alias match; no policy, history, model or incidents",
            "proposed": "Same text and catalog, simulated Maya profile, approved pre-existing synthetic articles/cases, deterministic policy; no active incident context in this text-only corpus",
        },
        "liveLLMBaseline": {
            "status": "skipped",
            "reason": "This command measures deterministic migration regression only and does not call a live model.",
        },
        "splits": {},
    }
    for split in ("dev", "heldout"):
        cases = [case for case in scenarios if case["split"] == split]
        result["splits"][split] = {}
        for method in ("keyword", "proposed"):
            rows = []
            for case in cases:
                start = time.perf_counter()
                # Pinned to v1: these published numbers are the scoring-v1 regression baseline
                # and must not move when the shipped default changes.
                d = decide(case["text"], sources, users[0], scoring="v1")
                keyword = next(
                    (
                        s
                        for s in catalog
                        if any(alias in case["text"].lower() for alias in s["aliases"])
                    ),
                    None,
                )
                team = (
                    (keyword["team"] if keyword else "Service Desk")
                    if method == "keyword"
                    else d["team"]
                )
                accepted = bool(keyword) if method == "keyword" else d["accepted"]
                rows.append(
                    {
                        "id": case["id"],
                        "family": case["family"],
                        "text": case["text"],
                        "expected": {
                            "team": case["team"],
                            "escalation": case["escalation"],
                            "clarification": case["expectedClarification"],
                        },
                        "actual": {
                            "team": team,
                            "accepted": accepted,
                            "escalation": "none" if method == "keyword" else d["escalation"],
                            "clarification": False if method == "keyword" else bool(d["question"]),
                            "reasons": ["first-keyword-match"]
                            if method == "keyword"
                            else d["reasons"],
                        },
                        "latencyMs": (time.perf_counter() - start) * 1000,
                    }
                )
            metrics = routing_metrics(rows)
            p50, p95 = percentiles([r["latencyMs"] for r in rows])
            result["splits"][split][method] = {
                **{key: value for key, value in metrics.items() if key != "failures"},
                "latency": {
                    "p50Ms": p50,
                    "p95Ms": p95,
                    "kind": "In-process deterministic route evaluation; includes proposed decision computation for both methods. Not UI or provider latency.",
                },
                "tokens": {"input": 0, "output": 0},
                "estimatedModelCostUSD": 0,
                "failures": metrics["failures"],
            }
    result["unmeasured"] = [
        "Live model performance/baseline/cost",
        "Human-reviewed label validity",
        "Related-incident precision/recall: text corpus lacks association labels",
        "Fact fidelity across broad language; targeted assertions exist in tests",
        "Real-world resolution effectiveness",
        "Browser and provider-confirmed latency",
    ]
    (root / "evaluation/python-results.json").write_text(json.dumps(result, indent=2) + "\n")
    lines = [
        "# Python migration regression results",
        "",
        f"Run: {result['runDate']}. Policy: {policy['version']}. Implementation SHA-256: {digest}.",
        "",
        "Labels are agent-authored and **not human reviewed**. Both historical split names are retained for comparison; their cases have already been observed. This is migration regression evidence, not a new untouched holdout or proof of improved real-world accuracy.",
        "",
        "| Split / method | Route accuracy | Accepted precision | Eligible coverage | Security recall |",
        "|---|---|---|---|---|",
    ]
    for split in ("dev", "heldout"):
        for method in ("keyword", "proposed"):
            metrics = result["splits"][split][method]
            values = " | ".join(
                _metric(metrics[key])
                for key in (
                    "routingAccuracy",
                    "acceptedPrecision",
                    "eligibleCoverage",
                    "securityRecall",
                )
            )
            lines.append(f"| {split} / {method} | {values} |")
    lines.extend(
        [
            "",
            "All failure cases and per-team counts are recorded in python-results.json. Historical TypeScript frozen.json and results.json remain unchanged. No live model calls occur; zero model cost is not an estimate of live costs.",
            "",
            "Limitations: " + "; ".join(result["unmeasured"]) + ".",
            "",
        ]
    )
    markdown = "\n".join(lines)
    (root / "evaluation/python-RESULTS.md").write_text(markdown)
    print(markdown)
    return result
